components/tidb_query_vec_expr/src/impl_string.rs

```diff
@@ -324,6 +324,38 @@ pub fn right(lhs: Option<BytesRef>, rhs: Option<&Int>) -> Result<Option<Bytes>>
     }
 }
 
+#[rpn_fn]
+#[inline]
+pub fn insert(
+    s: Option<BytesRef>,
+    pos: Option<&Int>,
+    len: Option<&Int>,
+    newstr: Option<BytesRef>,
+) -> Result<Option<Bytes>> {
+    match (s, pos, len, newstr) {
+        (Some(s), Some(pos), Some(len), Some(newstr)) => {
+            let pos = *pos;
+            let len = *len;
+            let upos: usize = pos as usize;
+            let mut ulen: usize = len as usize;
+            let s = s.to_vec();
```

Comment:
BytesRef's definition is `pub type BytesRef<'a> = &'a [u8];` which means we could eliminate this heap allocation here.